### ISSUE TITLE
Could not include required dependency warning

### ISSUE THREAD

**Issue by joeduffy** [pulumi/pulumi-aws#260]
I receive the warning

```
info: Could not include required dependency '@slack/client' in '/Users/joeduffy/temp/post-sqs-slack'.
```

for the below lambda.  I can't figure out why it's missing it, since it definitely does exist.  Is there a bug in the way we regex match scoped packages, because they exist inside of subdirectories?

```typescript
import * as pulumi from "@pulumi/pulumi";
import * as aws from "@pulumi/aws";
import * as serverless from "@pulumi/aws-serverless";

let config = new pulumi.Config(pulumi.getProject());
let queue = aws.sqs.Queue.get("my-queue", config.require("queueName"));
let slackChannel = config.get("slackChannel") || "#general";
let slackToken = config.require("slackToken");
serverless.queue.subscribe("my-slack-poster", queue, async (e: serverless.queue.QueueEvent) => {
    const slack = await import("@slack/client");
    const client = new slack.WebClient(slackToken);
    for (let rec of e.Records) {
        await client.chat.postMessage({
            channel: slackChannel,
            text: `*SQS message ${rec.messageId}* in region ${rec.awsRegion}:\n`+
                `${rec.body}\n`+
                `(with :love_letter: from Pulumi)`,
            as_user: true,
        });
    }
});
```

**Comment by joeduffy** [comment]
I suspect https://github.com/pulumi/pulumi-aws/blob/7786683cdeb1679cacd9752b4b88a6ebc7d2caf8/overlays/nodejs/serverless/function.ts#L315 is the culprit here, but I really don't follow what that code is trying to do.

Attempting a fix now, since this is blocking the demo I had wanted to do, ... please shout if anybody sees or knows anything about the underlying issue here.

**Comment by joeduffy** [comment]
Yep ☹️ 

Notice the leading slash in `/\/\@[^\/]*\/[^\/]*$/.exec(child.path)`.  This turns names like `@protobufjs/fetch` into `/@protobufjs/fetch`, which of course will not match the desired package name.

Simple one-line fix, which I have locally, and will push.  We need tests.

**Comment by lukehoban** [comment]
Argh: https://github.com/pulumi/pulumi-aws/pull/230#discussion_r193900791

The attempt to address that feedback led to this incorrect leading slash in the regexp.  I believe it works for nested dependencies in organizations, but not for top level dependencies in organizations.

> We need tests.

:100:

**Comment by joeduffy** [comment]
Hmm, well, the reason here is that `@slack/client` itself depends on the `loglevel` module, which has been installed at the top level `node_modules/` directory.  So, we have

```
node_modules/
    @slack/
        client/
            node_modules/
                <lots of stuff>
    loglevel/
```

It appears we happily zip up the `@slack/...` directory, but miss `loglevel`, and based on our current logic for doing this, this would seem to be "by design"?

Also, unless I'm missing something, I don't have a way of controlling what gets included in the functions created by the `aws-serverless` package, like I do `cloud-aws`?

Ultimately, I suspect https://github.com/pulumi/pulumi-aws/issues/257 is our way out here, but I seem to be somewhat in a pinch at the moment ...  I can try manually hacking some of my local packages.

**Comment by lukehoban** [comment]
> It appears we happily zip up the @slack/... directory, but miss loglevel, and based on our current logic for doing this, this would seem to be "by design"?

I can't reproduce this.  When I reference `@slack/client`, I see `loglevel` picked up and included in exactly the expected place in the `node_modules` tree.  The existing logic handles this case of dependencies which end up being placed higher in the tree by npm/yarn.

Can you share the ZIP for you Lambda?

**Comment by lukehoban** [comment]
> Also, unless I'm missing something, I don't have a way of controlling what gets included in the functions created by the aws-serverless package, like I do cloud-aws?

> Ultimately, I suspect #257 is our way out here, but I seem to be somewhat in a pinch at the moment ... I can try manually hacking some of my local packages.

Both true - #257 should remove the need for any manual overrides.  There may still be the need for additional user blacklisting (for performance, not correctness) which we'll need to expose config for at various layers.

**Comment by lukehoban** [comment]
> I can't reproduce this. When I reference @slack/client, I see loglevel picked up and included in exactly the expected place in the node_modules tree. The existing logic handles this case of dependencies which end up being placed higher in the tree by npm/yarn.

Indeed - it looks like this does reproduce in your branch for https://github.com/pulumi/pulumi-aws/pull/261, but does not reproduce when the one character fix (to remove the leading `/`) is made instead.  I'll need to look into why that is.  Currently testing a more targeted fix in #262 (including a test case for `@slack/client`).

**Comment by lukehoban** [comment]
Fixed with #262.
